start.py

```diff
@@ -1,47 +1,230 @@
-def print_board(board):
-    print("-------------")
-    for row in board:
-        for elem in row:
-            print(elem, end=' ')
-        print("\n-------------")
+import random
 
+class Player:
+    def __init__(self, name):
+        self._name = name
+        self._board = None
 
-def print_available_pieces(pieces):
-    print("Available Pieces:")
-    sorted_pieces = sorted(pieces)
-    for piece in sorted_pieces:
-        print(piece, end=" ")
+    def join_board(self, board):
+        self._board = board
 
+    def place(self, piece):
+        print("Otrzymana figura: \n%s" % piece)
+        while True:
+            row = input("Wybierz rząd (1-4): ")
+            while not (row.isnumeric() and 1 <= int(row) <= 4):
+                row = input("Wybrany rząd nie istnieje, ponów wybór: ")
+            row = int(row)
+            
+            col = input("Wybierz kolumnę (1-4): ")
+            while not (col.isnumeric() and 1 <= int(col) <= 4):
+                col = input("Wybrana kolumna nie istnieje, ponów wybór: ")
+            col = int(col)
+            
+            if self._board.is_avaliable(row, col):
+                self._board.place(piece, row, col)
+                quarto_call = input("Czy chcesz zawołać quatro? Wybierz TAK/NIE [t/n]: ")
+                while not quarto_call.upper() in ("T", "TAK", "N", "NIE"):
+                    quarto_call = input("Odpowiedź musi być w formacie TAK lub NIE [t/n]: ")
+                if quarto_call.upper() in ("T", "TAK"):
+                    if self._board.is_winning(piece):
+                        return True
+                    print("Podana figura nie tworzy Quatro")
+                return False
+            
+            print("Wybrane pole jest zajęte, wybierz inne.")
```

Comment:
to wszystko powyżej(czyli metodę place) można uprościć, dodając metodę do walidacji do tej klasy lub do innych. np. zrobić:

+ Piece.is_valid(piece)

i

+ tak samo w klasie Board

+ i można też sprawdzać wprowadzone dane w klasie Game

Reply:
Cytowany fragment jest odrobinę za krótki, ponieważ walidacja tak naprawdę się odbywa przez metodę Board.is_winning(Piece). W każdym razie w odpowiedzi [komentarz](https://github.com/wikimody/Quarto-project/pull/2#issuecomment-1843002606) zgodziłem się z poprzednimi spostrzeżeniami, co sprawiłoby, że cała metoda powinna zostać usunięta, jednak proponuję zastosować tu wymienioną nazwę "is_valid()" zamiast używanego dotychczas "is_winning()"

Reply:
chodziło mi o walidacje wprowadzonych przez użetkownika danych

Reply:
Wówczas Plansza, Gracz bądź inne klasy musiałby sprawdzać też niemożliwe przypadki typu nazwy, gdy obsługiwane są liczby. To raczej nieintuicyjne, dlatego proponuję pozostawić walidację w momencie przyjmowania danych lub oddelegować to do osobnej, dedykowanej klasy.

Reply:
lub rzuca wyjątek, wyjątek może nawet lepiej

Reply:
Moim zdaniem obsługa wyjątków lepiej się do tego nadaje, jednak spodobał mi się pomysł osobnej klasy (przykładowo Console) przejmującej w całości input i output:
Przykładowo wybór figury mógłby wyglądać następująco:
piece = self.find_piece(input.num("Wybierz figurę dla przeciwnika", self.avaliable_pieces_values()))